test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeTests.java

```diff
@@ -362,7 +362,60 @@ static public void testWithFieldOnNonExistentClass() throws Throwable {
 			Assert.fail("should throw error. Class does not exist");
 		} catch (NoClassDefFoundError e) {}
 	}
-	
+
+	/*
+	 * Test ifacmp on value class
+	 * 
+	 * class TestIfacmpOnValueClass {
+	 *  
+	 * }
+	 */
+	@Test(priority=1)
+	static public void TestIfacmpOnValueClass() throws Throwable {
+		String fields[] = {"x:I", "y:I"};
+		point2DClass = ValueTypeGenerator.generateValueClass("Point2DInAcmp", fields);
+		makePoint2D = lookup.findStatic(point2DClass, "makeValue", MethodType.methodType(point2DClass, int.class, int.class));
+
+		int x = 0;
+		int y = 0;
+		
+		Object valueType = makePoint2D.invoke(x, y);
+
+		Object refType = (Object) x;
+		
+		if (valueType == refType) {
+			Assert.fail("An identity (==) comparison that contains a valueType should always return false");
+		}
+		
+		if (refType == valueType) {
+			Assert.fail("An identity (==) comparison that contains a valueType should always return false");
+		}
+		
+		if (valueType == valueType) {
+			Assert.fail("An identity (==) comparison that contains a valueType should always return false");
+		}
+		
+		if (!(refType == refType)) {
+			Assert.fail("An identity (==) comparison on the same refType should always return true");
+		}
+        
+		if (!(valueType != refType)) {
```

Comment:
you need to double check that javac will not optimize `(!(valueType != refType))` into `(valueType == refType)`

Reply:
``if (false == (valueType != refType))``
Does java optimize this code? I think it may not.

Reply:
you can check by using javap with the -c option to see what bytecodes the javac compiler generated